### PR TITLE
Improvements in maplab_node synchronizer/data handling

### PR DESCRIPTION
**Pull request by mfehr**
### Changes

- Adds option to batch IMU measurements to reduce overhead of handing
high frequency data.
```yaml
# To reduce the cost of spawning threads and processing each imu measurement
# individually, the IMU messages can be batched. Setting this frequency to <=0 disables batching.
maplab_batch_imu_measurements_at_frequency: 20
```
- Adds option to throttle external odometry data (same reason, e.g. MSF
publishes at IMU frequency)
```yaml
# If the odometry fed to the maplab_node has a high frequency, throttling is adviced.
maplab_throttle_frequency_odometry: 20
```
- Fix conceptual error in synchronizer. Before reaching the
synchronizer, the images were processed by the visual pipeline (create
multi-cam frame, feature tracking) to create the VisualNFrames. Only
then the synchronizer would hold those back until there is enough IMU
and odometry data available. However this is too late, since feature
tracking already needs IMU data to work properly. In some cases feature
tracking failed due to this error.

### Results
**Without batching/throttling, running a 220 Hz odom and imu topic (100% = 1 thread)**
1583_without_throttling
**With the above settings (100% = 1 thread)**
27295_w_throttling

### CODE REVIEW

**Comment by floriantschopp** [comment]
test this please

**Comment by smauq** [comment]
test this please

File: applications/maplab-node/src/synchronizer.cc [hunk 1]
```diff
@@ -323,34 +333,92 @@ void Synchronizer::processOdometryMeasurement(
...
+  // Release all visual frames that are ready.
+  Aligned<std::vector, aslam::VisualNFrame::Ptr> finished_nframes;
+  aslam::VisualNFrame::Ptr next_nframe = visual_pipeline_->getNext();
+  size_t num_nframes_released = 0u;
+  while (next_nframe) {
+    const int64_t current_frame_timestamp_ns =
+        next_nframe->getMinTimestampNanoseconds();
 
-    frame_skip_counter_ += dropped_nframes;
+    VLOG(5) << "[MaplabNode-Synchronizer] retrieved finished VisualNFrame at "
```
Comment (LBern) [inline applications/maplab-node/src/synchronizer.cc:406]: minor: if verbosity is < 5, `current_frame_timestamp_ns` is not used in this loop

File: applications/maplab-node/src/ros-helpers.cc [hunk 2]
```diff
@@ -42,16 +42,16 @@ DEFINE_bool(
...
-  vio::ImuMeasurement::Ptr imu_measurement(new vio::ImuMeasurement);
-  imu_measurement->timestamp = rosTimeToNanoseconds(imu_msg->header.stamp);
-  imu_measurement->imu_data << imu_msg->linear_acceleration.x,
-      imu_msg->linear_acceleration.y, imu_msg->linear_acceleration.z,
-      imu_msg->angular_velocity.x, imu_msg->angular_velocity.y,
-      imu_msg->angular_velocity.z;
-  return imu_measurement;
+void addRosImuMeasurementToImuMeasurementBatch(
+    const sensor_msgs::Imu& imu_msg,
+    vio::BatchedImuMeasurements* batched_imu_measurements_ptr) {
```
Comment (LBern) [inline applications/maplab-node/src/ros-helpers.cc:47]: Nice! This avoids any copy/move construction of the new IMU measurement

**Comment by smauq** [comment]
Seems that for some reason this doesn't consistently pass the end to end test. Seems to be a problem due to the trajectory alignment not working, maybe this introduces some sort of different time stamps on the vertices? I think we still need to debug what the issue is as I feel there is still something.

**Comment by LBern** [comment]
@smauq There was an issue in the odometry throttler which ended up discarding essentially all odometry measurements and thus, the end-to-end tests were failing.

**Comment by floriantschopp** [comment]
Can we merge this or does it require more review?

**Comment by LBern** [comment]
Let's merge it

**Pull request merged**